File: sdk/typescript/src/contracts/policy.ts
import { Keypair } from "@stellar/stellar-sdk";
import {
  SorobanClient,
  TxOptions,
  toAddress,
  toI128,
  toU64,
  toU32,
  toString,
  toBool,
  fromScVal,
} from "../client";
import {
  InitializePolicySchema,
  IssuePolicySchema,
  IssuePolicyWithDurationSchema,
  CancelPolicySchema,
  ExpirePolicySchema,
  CheckExpirePoliciesSchema,
  CalculateDynamicPremiumSchema,
  CreateAmmPoolSchema,
  SwapPolicyFractionSchema,
  RequestEndorsementSchema,
  ApproveEndorsementSchema,
  RejectEndorsementSchema,
  SetPauseStateSchema,
  type PolicyRecord,
} from "../schemas";

export class PolicyContract {
  constructor(
    private readonly client: SorobanClient,
    private readonly contractId: string
  ) {}

  async initialize(
    admin: string,
    guardian: string,
    opts: TxOptions
  ): Promise<void> {
    const params = InitializePolicySchema.parse({ admin, guardian });
    await this.client.invoke(
      this.contractId,
      "initialize",
      [toAddress(params.admin), toAddress(params.guardian)],
      opts
    );
  }

  async issuePolicy(
    holder: string,
    policyId: bigint,
    coverage: bigint,
    premium: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = IssuePolicySchema.parse({
      holder,
      policy_id: policyId,
      coverage,
      premium,
    });
    await this.client.invoke(
      this.contractId,
      "issue_policy",
      [
        toAddress(params.holder),
        toI128(params.policy_id),
        toI128(params.coverage),
        toI128(params.premium),
      ],
      opts
    );
  }

  async issuePolicyWithDuration(
    holder: string,
    policyId: bigint,
    coverage: bigint,
    premium: bigint,
    durationDays: number,
    opts: TxOptions
  ): Promise<void> {
    const params = IssuePolicyWithDurationSchema.parse({
      holder,
      policy_id: policyId,
      coverage,
      premium,
      duration_days: durationDays,
    });
    await this.client.invoke(
      this.contractId,
      "issue_policy_with_duration",
      [
        toAddress(params.holder),
        toI128(params.policy_id),
        toI128(params.coverage),
        toI128(params.premium),
        toU32(params.duration_days),
      ],
      opts
    );
  }

  async cancelPolicy(
    policyId: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = CancelPolicySchema.parse({ policy_id: policyId });
    await this.client.invoke(
      this.contractId,
      "cancel_policy",
      [toI128(params.policy_id)],
      opts
    );
  }

  async expirePolicy(
    policyId: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = ExpirePolicySchema.parse({ policy_id: policyId });
    await this.client.invoke(
      this.contractId,
      "expire_policy",
      [toI128(params.policy_id)],
      opts
    );
  }

  async checkAndExpirePolicies(
    startIndex: bigint,
    maxItems: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = CheckExpirePoliciesSchema.parse({
      start_index: startIndex,
      max_items: maxItems,
    });
    await this.client.invoke(
      this.contractId,
      "check_and_expire_policies",
      [toI128(params.start_index), toI128(params.max_items)],
      opts
    );
  }

  async isPolicyActive(
    policyId: bigint,
    opts: TxOptions
  ): Promise<boolean> {
    const result = await this.client.invoke(
      this.contractId,
      "is_policy_active",
      [toI128(policyId)],
      { ...opts, simulate: true }
    );
    return fromScVal<boolean>(result);
  }

  async getPolicyCoverage(
    policyId: bigint,
    opts: TxOptions
  ): Promise<bigint> {
    const result = await this.client.invoke(
      this.contractId,
      "get_policy_coverage",
      [toI128(policyId)],
      { ...opts, simulate: true }
    );
    return fromScVal<bigint>(result);
  }

  async calculateDynamicPremium(
    basePremium: bigint,
    riskScore: number,
    locationFactor: number,
    coverageRatio: number,
    opts: TxOptions
  ): Promise<bigint> {
    const params = CalculateDynamicPremiumSchema.parse({
      base_premium: basePremium,
      risk_score: riskScore,
      location_factor: locationFactor,
      coverage_ratio: coverageRatio,
    });
    const result = await this.client.invoke(
      this.contractId,
      "calculate_dynamic_premium",
      [
        toI128(params.base_premium),
        toU32(params.risk_score),
        toU32(params.location_factor),
        toU32(params.coverage_ratio),
      ],
      { ...opts, simulate: true }
    );
    return fromScVal<bigint>(result);
  }

  async createAmmPool(
    policyId: bigint,
    stableAmount: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = CreateAmmPoolSchema.parse({
      policy_id: policyId,
      stable_amount: stableAmount,
    });
    await this.client.invoke(
      this.contractId,
      "create_amm_pool",
      [toI128(params.policy_id), toI128(params.stable_amount)],
      opts
    );
  }

  async swapPolicyFraction(
    buyer: string,
    policyId: bigint,
    stableIn: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = SwapPolicyFractionSchema.parse({
      buyer,
      policy_id: policyId,
      stable_in: stableIn,
    });
    await this.client.invoke(
      this.contractId,
      "swap_policy_fraction",
      [
        toAddress(params.buyer),
        toI128(params.policy_id),
        toI128(params.stable_in),
      ],
      opts
    );
  }

  async requestEndorsement(
    requester: string,
    policyId: bigint,
    newCoverage: bigint,
    newPremium: bigint,
    reason: string,
    opts: TxOptions
  ): Promise<void> {
    const params = RequestEndorsementSchema.parse({
      requester,
      policy_id: policyId,
      new_coverage: newCoverage,
      new_premium: newPremium,
      reason,
    });
    await this.client.invoke(
      this.contractId,
      "request_endorsement",
      [
        toAddress(params.requester),
        toI128(params.policy_id),
        toI128(params.new_coverage),
        toI128(params.new_premium),
        toString(params.reason),
      ],
      opts
    );
  }

  async approveEndorsement(
    caller: string,
    endorsementId: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = ApproveEndorsementSchema.parse({
      caller,
      endorsement_id: endorsementId,
    });
    await this.client.invoke(
      this.contractId,
      "approve_endorsement",
      [toAddress(params.caller), toI128(params.endorsement_id)],
      opts
    );
  }

  async rejectEndorsement(
    caller: string,
    endorsementId: bigint,
    reason: string,
    opts: TxOptions
  ): Promise<void> {
    const params = RejectEndorsementSchema.parse({
      caller,
      endorsement_id: endorsementId,
      reason,
    });
    await this.client.invoke(
      this.contractId,
      "reject_endorsement",
      [
        toAddress(params.caller),
        toI128(params.endorsement_id),
        toString(params.reason),
      ],
      opts
    );
  }

  async setPauseState(
    caller: string,
    isPaused: boolean,
    reason: string,
    opts: TxOptions
  ): Promise<void> {
    const params = SetPauseStateSchema.parse({
      caller,
      is_paused: isPaused,
      reason,
    });
    await this.client.invoke(
      this.contractId,
      "set_pause_state",
      [toAddress(params.caller), toBool(params.is_paused), toString(params.reason)],
      opts
    );
  }
}

File: sdk/python/stellar_insured_sdk/contracts/policy.py
"""Policy contract wrapper with schema-validated payloads."""

from __future__ import annotations

from ..client import (
    SorobanClient,
    TxOptions,
    to_address,
    to_int128,
    to_uint32,
    to_bool,
    to_string,
    from_scval,
)
from ..schemas import (
    InitializePolicyParams,
    IssuePolicyParams,
    IssuePolicyWithDurationParams,
    CalculateDynamicPremiumParams,
    CreateAmmPoolParams,
    SwapPolicyFractionParams,
    RequestEndorsementParams,
    ApproveEndorsementParams,
    RejectEndorsementParams,
    SetPauseStateParams,
    PolicyRecord,
)


class PolicyContract:
    """Typed wrapper for the Policy Soroban contract."""

    def __init__(self, client: SorobanClient, contract_id: str) -> None:
        self._client = client
        self._contract_id = contract_id

    def initialize(self, admin: str, guardian: str, opts: TxOptions) -> None:
        params = InitializePolicyParams(admin=admin, guardian=guardian)
        self._client.invoke(
            self._contract_id,
            "initialize",
            [to_address(params.admin), to_address(params.guardian)],
            opts,
        )

    def issue_policy(
        self,
        holder: str,
        policy_id: int,
        coverage: int,
        premium: int,
        opts: TxOptions,
    ) -> None:
        params = IssuePolicyParams(
            holder=holder, policy_id=policy_id, coverage=coverage, premium=premium
        )
        self._client.invoke(
            self._contract_id,
            "issue_policy",
            [
                to_address(params.holder),
                to_int128(params.policy_id),
                to_int128(params.coverage),
                to_int128(params.premium),
            ],
            opts,
        )

    def issue_policy_with_duration(
        self,
        holder: str,
        policy_id: int,
        coverage: int,
        premium: int,
        duration_days: int,
        opts: TxOptions,
    ) -> None:
        params = IssuePolicyWithDurationParams(
            holder=holder,
            policy_id=policy_id,
            coverage=coverage,
            premium=premium,
            duration_days=duration_days,
        )
        self._client.invoke(
            self._contract_id,
            "issue_policy_with_duration",
            [
                to_address(params.holder),
                to_int128(params.policy_id),
                to_int128(params.coverage),
                to_int128(params.premium),
                to_uint32(params.duration_days),
            ],
            opts,
        )

    def cancel_policy(self, policy_id: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id, "cancel_policy", [to_int128(policy_id)], opts
        )

    def expire_policy(self, policy_id: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id, "expire_policy", [to_int128(policy_id)], opts
        )

    def check_and_expire_policies(
        self, start_index: int, max_items: int, opts: TxOptions
    ) -> None:
        self._client.invoke(
            self._contract_id,
            "check_and_expire_policies",
            [to_int128(start_index), to_int128(max_items)],
            opts,
        )

    def is_policy_active(self, policy_id: int, opts: TxOptions) -> bool:
        result = self._client.invoke(
            self._contract_id,
            "is_policy_active",
            [to_int128(policy_id)],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def get_policy_coverage(self, policy_id: int, opts: TxOptions) -> int:
        result = self._client.invoke(
            self._contract_id,
            "get_policy_coverage",
            [to_int128(policy_id)],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def calculate_dynamic_premium(
        self,
        base_premium: int,
        risk_score: int,
        location_factor: int,
        coverage_ratio: int,
        opts: TxOptions,
    ) -> int:
        params = CalculateDynamicPremiumParams(
            base_premium=base_premium,
            risk_score=risk_score,
            location_factor=location_factor,
            coverage_ratio=coverage_ratio,
        )
        result = self._client.invoke(
            self._contract_id,
            "calculate_dynamic_premium",
            [
                to_int128(params.base_premium),
                to_uint32(params.risk_score),
                to_uint32(params.location_factor),
                to_uint32(params.coverage_ratio),
            ],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def create_amm_pool(
        self, policy_id: int, stable_amount: int, opts: TxOptions
    ) -> None:
        params = CreateAmmPoolParams(policy_id=policy_id, stable_amount=stable_amount)
        self._client.invoke(
            self._contract_id,
            "create_amm_pool",
            [to_int128(params.policy_id), to_int128(params.stable_amount)],
            opts,
        )

    def swap_policy_fraction(
        self, buyer: str, policy_id: int, stable_in: int, opts: TxOptions
    ) -> None:
        params = SwapPolicyFractionParams(
            buyer=buyer, policy_id=policy_id, stable_in=stable_in
        )
        self._client.invoke(
            self._contract_id,
            "swap_policy_fraction",
            [
                to_address(params.buyer),
                to_int128(params.policy_id),
                to_int128(params.stable_in),
            ],
            opts,
        )

    def request_endorsement(
        self,
        requester: str,
        policy_id: int,
        new_coverage: int,
        new_premium: int,
        reason: str,
        opts: TxOptions,
    ) -> None:
        params = RequestEndorsementParams(
            requester=requester,
            policy_id=policy_id,
            new_coverage=new_coverage,
            new_premium=new_premium,
            reason=reason,
        )
        self._client.invoke(
            self._contract_id,
            "request_endorsement",
            [
                to_address(params.requester),
                to_int128(params.policy_id),
                to_int128(params.new_coverage),
                to_int128(params.new_premium),
                to_string(params.reason),
            ],
            opts,
        )

    def approve_endorsement(
        self, caller: str, endorsement_id: int, opts: TxOptions
    ) -> None:
        params = ApproveEndorsementParams(
            caller=caller, endorsement_id=endorsement_id
        )
        self._client.invoke(
            self._contract_id,
            "approve_endorsement",
            [to_address(params.caller), to_int128(params.endorsement_id)],
            opts,
        )

    def reject_endorsement(
        self, caller: str, endorsement_id: int, reason: str, opts: TxOptions
    ) -> None:
        params = RejectEndorsementParams(
            caller=caller, endorsement_id=endorsement_id, reason=reason
        )
        self._client.invoke(
            self._contract_id,
            "reject_endorsement",
            [
                to_address(params.caller),
                to_int128(params.endorsement_id),
                to_string(params.reason),
            ],
            opts,
        )

    def set_pause_state(
        self, caller: str, is_paused: bool, reason: str, opts: TxOptions
    ) -> None:
        params = SetPauseStateParams(
            caller=caller, is_paused=is_paused, reason=reason
        )
        self._client.invoke(
            self._contract_id,
            "set_pause_state",
            [
                to_address(params.caller),
                to_bool(params.is_paused),
                to_string(params.reason),
            ],
            opts,
        )
